src/blueapi/core/context.py

```diff
@@ -144,20 +146,41 @@ def device(self, device: Device, name: Optional[str] = None) -> None:
             raise TypeError(f"{device} is not a Bluesky compatible device")
 
         if name is None:
-            if isinstance(device, Readable) or isinstance(device, Flyable):
+            if isinstance(device, HasName):
                 name = device.name
             else:
                 raise KeyError(f"Must supply a name for this device: {device}")
 
         self.devices[name] = device
 
 
-def device_validators(ctx: BlueskyContext) -> Iterable[TypeValidatorDefinition]:
-    def get_device(name: str) -> Device:
-        device = ctx.find_device(name)
-        if device is None:
-            raise KeyError(f"Could not find a device named {name}")
-        return device
-
-    for proto in BLUESKY_PROTOCOLS:
-        yield TypeValidatorDefinition(proto, get_device)
+def plan_validators(
+    ctx: BlueskyContext, plan: PlanGenerator
+) -> Iterable[TypeValidatorDefinition]:
+    def get_validator(plan_arg: str, field_type: Type) -> TypeValidatorDefinition:
+        def get_device(name: str):
+            device = ctx.find_device(name)
+            if device is None:
+                raise KeyError(
+                    f"Could not find a device named {name} for plan argument {plan_arg}"
+                )
+            if not isinstance(device, field_type):
+                raise KeyError(
+                    f"Device named {name} did not comply with expected type"
+                    f" {field_type} of plan argument {plan_arg}"
+                )
+            return device
+
+        return TypeValidatorDefinition(field_type, get_device)
+
+    def get_validators(
+        plan_arg: str, field_type: Type
+    ) -> Iterable[TypeValidatorDefinition]:
+        if isinstance(field_type, _GenericAlias):
+            for arg in get_args(field_type):
+                yield from get_validators(plan_arg, arg)
+        elif is_bluesky_compatible_device_type(field_type):
+            yield get_validator(plan_arg, field_type)
+
+    for k, v in Signature.from_callable(plan).parameters.items():
```

Comment:
This will fail if a plan has two arguments of the same type because `utils.type_validator` doesn't allow it. Example test case:
```python
def concrete_plan_multi_arg(
    motor_1: DelegatingNamedMovable,
    motor_2: DelegatingNamedMovable,
) -> MsgGenerator:  # type: ignore
    ...
```
Presumably some sort of caching is needed